File: code/packages/fsharp/type-checker-protocol/TypeCheckerProtocol.fs
namespace CodingAdventures.TypeCheckerProtocol.FSharp

open System
open System.Collections.Generic
open System.Text.RegularExpressions

type TypeErrorDiagnostic =
    { Message: string
      Line: int
      Column: int }

type TypeCheckResult<'Ast> =
    { TypedAst: 'Ast
      Errors: TypeErrorDiagnostic list }

    member this.Ok = List.isEmpty this.Errors

type ITypeChecker<'AstIn, 'AstOut> =
    abstract Check: 'AstIn -> TypeCheckResult<'AstOut>

[<Sealed>]
type NotHandled private () =
    static member val Value = NotHandled()

[<AbstractClass>]
type GenericTypeChecker<'Ast>() =
    let errors = ResizeArray<TypeErrorDiagnostic>()
    let hooks = Dictionary<string, ResizeArray<'Ast -> obj array -> obj>>()

    static member NormalizeKind(kind: string option) =
        match kind with
        | None
        | Some "" -> ""
        | Some value -> Regex.Replace(value, @"\W+", "_").Trim('_')

    member _.RegisterHook(phase: string, kind: string, hook: 'Ast -> obj array -> obj) =
        if isNull phase then nullArg "phase"
        if isNull kind then nullArg "kind"
        if isNull (box hook) then nullArg "hook"

        let key = $"{phase}:{GenericTypeChecker<'Ast>.NormalizeKind(Some kind)}"

        let bucket =
            match hooks.TryGetValue key with
            | true, existing -> existing
            | false, _ ->
                let created = ResizeArray()
                hooks[key] <- created
                created

        bucket.Add hook

    member this.Dispatch(phase: string, node: 'Ast, ?args: obj array, ?defaultValue: obj) =
        if isNull phase then nullArg "phase"

        let args = defaultArg args [||]
        let defaultValue = defaultArg defaultValue null
        let normalized = this.NodeKind node |> GenericTypeChecker<'Ast>.NormalizeKind
        let keys = [ $"{phase}:{normalized}"; $"{phase}:*" ]
        let mutable handled = false
        let mutable output = defaultValue

        for key in keys do
            if not handled then
                match hooks.TryGetValue key with
                | true, bucket ->
                    for hook in bucket do
                        if not handled then
                            let result = hook node args
                            if not (Object.ReferenceEquals(result, NotHandled.Value)) then
                                output <- result
                                handled <- true
                | false, _ -> ()

        output

    member this.Check(ast: 'Ast) =
        errors.Clear()
        this.Run ast

        { TypedAst = ast
          Errors = errors |> Seq.toList }

    member this.Error(message: string, subject: obj) =
        let line, column = this.Locate subject
        errors.Add({ Message = message; Line = line; Column = column })

    abstract Run: 'Ast -> unit
    abstract NodeKind: 'Ast -> string option
    abstract Locate: obj -> int * int
    default _.Locate(_subject: obj) = 1, 1

    interface ITypeChecker<'Ast, 'Ast> with
        member this.Check ast = this.Check ast

[<RequireQualifiedAccess>]
module TypeCheckerProtocolPackage =
    [<Literal>]
    let Version = "0.1.0"

File: code/packages/go/type-checker-protocol/type_checker_protocol.go
package typecheckerprotocol

import "strings"

type TypeErrorDiagnostic struct {
	Message string
	Line    int
	Column  int
}

type TypeCheckResult[T any] struct {
	TypedAST T
	Errors   []TypeErrorDiagnostic
	OK       bool
}

type TypeChecker[ASTIn any, ASTOut any] interface {
	Check(ast ASTIn) TypeCheckResult[ASTOut]
}

type HookResult any
type Hook[T any] func(node T, args ...any) HookResult
type Locator func(subject any) (int, int)
type NodeKind[T any] func(node T) string

type notHandledToken struct{}

type GenericTypeChecker[T any] struct {
	hooks    map[string][]Hook[T]
	errors   []TypeErrorDiagnostic
	nodeKind NodeKind[T]
	locate   Locator
}

func NewGenericTypeChecker[T any](nodeKind NodeKind[T], locate Locator) *GenericTypeChecker[T] {
	if locate == nil {
		locate = func(subject any) (int, int) {
			_ = subject
			return 1, 1
		}
	}

	return &GenericTypeChecker[T]{
		hooks:    map[string][]Hook[T]{},
		nodeKind: nodeKind,
		locate:   locate,
	}
}

func (g *GenericTypeChecker[T]) Reset() {
	g.errors = nil
}

func (g *GenericTypeChecker[T]) RegisterHook(phase string, kind string, hook Hook[T]) {
	key := phase + ":" + normalizeKind(kind)
	g.hooks[key] = append(g.hooks[key], hook)
}

func (g *GenericTypeChecker[T]) Dispatch(phase string, node T, args ...any) HookResult {
	kind := ""
	if g.nodeKind != nil {
		kind = normalizeKind(g.nodeKind(node))
	}

	keys := []string{phase + ":" + kind, phase + ":*"}
	for _, key := range keys {
		for _, hook := range g.hooks[key] {
			result := hook(node, args...)
			if _, ok := result.(notHandledToken); !ok {
				return result
			}
		}
	}

	return nil
}

func (g *GenericTypeChecker[T]) NotHandled() HookResult {
	return notHandledToken{}
}

func (g *GenericTypeChecker[T]) Error(message string, subject any) {
	line, column := g.locate(subject)
	g.errors = append(g.errors, TypeErrorDiagnostic{
		Message: message,
		Line:    line,
		Column:  column,
	})
}

func (g *GenericTypeChecker[T]) Errors() []TypeErrorDiagnostic {
	out := make([]TypeErrorDiagnostic, len(g.errors))
	copy(out, g.errors)
	return out
}

func normalizeKind(kind string) string {
	var builder strings.Builder
	lastUnderscore := false

	for _, char := range kind {
		if (char >= 'a' && char <= 'z') || (char >= 'A' && char <= 'Z') || (char >= '0' && char <= '9') {
			builder.WriteRune(char)
			lastUnderscore = false
			continue
		}

		if !lastUnderscore {
			builder.WriteByte('_')
			lastUnderscore = true
		}
	}

	return strings.Trim(builder.String(), "_")
}
